### PR TITLE
LUKS2 tests: generating test-containers on-the-fly.

### PR DESCRIPTION
**Pull request by thatux**
Background: For LUKS1 we have [pre-generated/static test-containers](https://hashcat.net/misc/example_hashes/hashcat_luks_testfiles.7z): one for each option (e.g. hashing- & crypto-algorithm), all using the password `hashcat`.

LUKS2 introduces more options (e.g. Argon2 parameters), this would require pre-generating many different containers, resulting in a large archive as the [default metadata size is 16MiB](https://gitlab.com/cryptsetup/cryptsetup/-/blob/main/docs/v2.1.0-ReleaseNotes#L27).
Still that would only test a single pre-generated password per container.

I've chosen to generate the LUKS2 tests-containers dynamically: m34100.pm calls m34100.sh with takes a password as argument and uses cryptsetup with random parameters to generate a containers.
It loop-mounts the container, creates a ext4 filesystem on it, and puts a text document in the container.

TODO:
- [x] Fix bug found by these tests: 64 (and some longer passwords lengths) for LUKS2 (and Argon2) don't crack --> https://github.com/hashcat/hashcat/pull/4447
- [x] finishes TODO in https://github.com/hashcat/hashcat/pull/4415
- [x] add all dependencies to install_modules.sh --> only cryptsetup !?
- [x] how to take sudo into account? --> only sudo where required & generated tests optional (using --generate flag)
- [x] do we need the md5 compare? --> I think so because of the large LUKS2 hashes (which cannot be passed on the commandline), if not now, then probably later..
- [x] add GENERATE_CONTAINERS_MODES array containing attacks that have dynamically generated tests (containers?) such that we can compare to GENERATE_CONTAINERS_MODES list of container-generating hashes (instead of only LUKS2)
- [x] currently only works for LUKS2 with -a0 --> expected behavior as it is part of SLOW_ALGOS and those have only a single kernel, therefore we don't need to test the other modes.
- [x]  test_modules/m34100.sh: do not use /tmp in m34100.sh, change OUTPUT_DIR and MOUNT_DIR to test-folder ($OUTD ) --> I'm fine with this
- [x] test.sh: try not to specifically match on '$luks$' --> don't care, we're making exceptions for LUKS anyways
- [x]  add static LUKS2 container tests: this approach is not crossplatform (windows/macos) which is why these tests are optional with -g flag. The generated-container tests are an addition to the static tests as it finds a different class of bugs (e.g. password lengths).

Future work:
- [x] support LUKS1 dynamic containers --> we can probably reuse (most of) m34100.sh but with luks1 as parameter to create_luks_container() --> https://github.com/hashcat/hashcat/pull/4487
- [ ] support GPG
- [ ] support PKZIP
- [ ]  test_modules/m34100.sh: generate multiple keyslots with random passwords, and one keyslot with the test.pl password, and pick that keyslot

Who wants to test `./tools/test.sh -m 34000-34300 -D1 -g` to check if this also works for you?

### CODE REVIEW

**Comment by thatux** [comment]
Maybe I'll try and add LUKS2 static tests in the future, anybody has a preference on the parameters tested and a maximum size for the download? [LUKS1](https://hashcat.net/misc/example_hashes/hashcat_luks_testfiles.7z) is ~40mb and tests 91 different options.

I've now generated 886 LUKS2 containers ~9gb with all different options (see attached), or we can limit to 147 luks2-**aes** options ~3gb [luks2_container_options.txt](https://github.com/user-attachments/files/21940021/luks2_container_options.txt)


But I opt to merge this PR first/anyway

**Comment by jsteube** [comment]
Should we really try Argon2 configurations? I fully understand all the Ciphers/Cascades.

**Comment by thatux** [comment]
Maybe not, probably the argon2 34000 test will catch most of the errors, but I'm not familiar enough with LUKS2 to be certain (don't know whether there is complexity in that part of the algorithm).

Just start without those?

**Comment by thatux** [comment]
cleanup & compile
```
make -s distclean; make -sj40
```

Static tests work :smiley: 
```
bash ./tools/test.sh -m 34100 -D1 -t all     
[ test_1757359359 ] > Init test for hash type 34100.

ATTENTION: the luks2 test files (for -m 34100) are currently missing on your system.
They will be fetched from https://github.com/thatux/hashcat/raw/refs/heads/luks2_test_clean/tools/luks2_tests.7z
Note: this needs to be done only once and could take a little bit to download/extract.
These luks2 test files are not shipped directly with hashcat because the file sizes are
particularly large and therefore a bandwidth burner for users who do not run these tests.

[ test_1757359359 ] [ Type 34100, Attack 0, Mode single, Device-Type Cpu, Kernel-Type Optimized, Vector-Width 1, LUKS2-mode luks2-aes-argon2id-t4-m16-p1-size20MiB ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1757359359 ] [ Type 34100, Attack 0, Mode single, Device-Type Cpu, Kernel-Type Optimized, Vector-Width 1, LUKS2-mode luks2-aes-argon2i-t4-m16-p1-size20MiB ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1757359359 ] [ Type 34100, Attack 0, Mode single, Device-Type Cpu, Kernel-Type Optimized, Vector-Width 1, LUKS2-mode luks2-aes-sha256-size20MiB ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1757359359 ] [ Type 34100, Attack 0, Mode single, Device-Type Cpu, Kernel-Type Optimized, Vector-Width 4, LUKS2-mode luks2-aes-argon2id-t4-m16-p1-size20MiB ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1757359359 ] [ Type 34100, Attack 0, Mode single, Device-Type Cpu, Kernel-Type Optimized, Vector-Width 4, LUKS2-mode luks2-aes-argon2i-t4-m16-p1-size20MiB ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1757359359 ] [ Type 34100, Attack 0, Mode single, Device-Type Cpu, Kernel-Type Optimized, Vector-Width 4, LUKS2-mode luks2-aes-sha256-size20MiB ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
```

Dynamic tests also work  :smiley: 
```
bash ./tools/test.sh -m 34100 -D1 -t all -g
We'll need sudo to generate crypto-containers on-the-fly
[sudo] password for user: 
[ test_1757359534 ] > Init test for hash type 34100.
[ test_1757359534 ] [ Type 34100, Attack 0, Mode single, Device-Type Cpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1757359534 ] [ Type 34100, Attack 0, Mode multi,  Device-Type Cpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1757359534 ] [ Type 34100, Attack 0, Mode single, Device-Type Cpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1757359534 ] [ Type 34100, Attack 0, Mode multi,  Device-Type Cpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
```

I thinks this is complete and can be merged

@jsteube do you want to host https://github.com/thatux/hashcat/raw/refs/heads/luks2_test_clean/tools/luks2_tests.7z on hashcat.net?

**Comment by jsteube** [comment]
Yes, let's host them on hashcat.net. Are they ready to be moved?

**Comment by thatux** [comment]
Yes, if you can upload the .7z, than I can edit the code to refer to that URL (tomorrow).

**Comment by jsteube** [comment]
Up at https://hashcat.net/misc/example_hashes/luks2_tests.7z

File: tools/README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+# Enable LUKS2 on-the-fly crypto-container generation
+sudo apt install cryptsetup
+
+./install_modules.sh
+```
+
+
+### Example usage
+```
+./test.sh -m 0 -D1 -t all
```
Comment (jsteube) [inline tools/README.md]: Requires the user to have a CPU runtime installed, which is rare. I'd say avoid -D1 at all

File: tools/test.sh [hunk 2]
```diff
@@ -248,6 +257,69 @@ function init()
...
+      mkdir -p "${luks2_tests_folder}"
+    fi
+
+    luks2_first_test_file="${luks2_tests_folder}/luks2-aes-argon2id-t4-m16-p1-size20MiB.img"
+
+    if [ ! -f "${luks2_first_test_file}" ]; then
+      # luks2_tests="hashcat_luks2_testfiles.7z"
+      # luks2_tests_url="https://hashcat.net/misc/example_hashes/${luks2_tests}"
+      luks2_tests="luks2_tests.7z"
+      luks2_tests_url="https://github.com/thatux/hashcat/raw/refs/heads/luks2_test_clean/tools/${luks2_tests}"
```
Comment (jsteube) [inline tools/test.sh]: up on https://hashcat.net/misc/example_hashes/luks2_tests.7z

File: tools/test_modules/m34100.sh [hunk 3]
```diff
@@ -0,0 +1,132 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+if [[ -z "${1:-}" ]]; then
+#   echo "❌ Missing argument: password"
```
Comment (jsteube) [inline tools/test_modules/m34100.sh]: Please get rid of this utf8 stuff :)

**Comment by jsteube** [comment]
Worked on first try!

```
atom@x870:~/hashcat$ tools/test.sh -m 34100
[ test_1757440063 ] > Init test for hash type 34100.

ATTENTION: the luks2 test files (for -m 34100) are currently missing on your system.
They will be fetched from https://github.com/thatux/hashcat/raw/refs/heads/luks2_test_clean/tools/luks2_tests.7z
Note: this needs to be done only once and could take a little bit to download/extract.
These luks2 test files are not shipped directly with hashcat because the file sizes are
particularly large and therefore a bandwidth burner for users who do not run these tests.

[ test_1757440063 ] [ Type 34100, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1, LUKS2-mode luks2-aes-argon2id-t4-m16-p1-size20MiB ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1757440063 ] [ Type 34100, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1, LUKS2-mode luks2-aes-argon2i-t4-m16-p1-size20MiB ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1757440063 ] [ Type 34100, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1, LUKS2-mode luks2-aes-sha256-size20MiB ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1757440063 ] [ Type 34100, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4, LUKS2-mode luks2-aes-argon2id-t4-m16-p1-size20MiB ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1757440063 ] [ Type 34100, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4, LUKS2-mode luks2-aes-argon2i-t4-m16-p1-size20MiB ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1757440063 ] [ Type 34100, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4, LUKS2-mode luks2-aes-sha256-size20MiB ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
```

Also I saw from the sources that you've added a "-g" mode:

```
atom@x870:~/hashcat$ tools/test.sh -m 34100 -g
We'll need sudo to generate crypto-containers on-the-fly
[ test_1757440277 ] > Init test for hash type 34100.
[ test_1757440277 ] [ Type 34100, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1757440277 ] [ Type 34100, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
```

File: tools/README.md (hunk 1 above)
Comment (thatux) [reply tools/README.md]: done: removed `-D1`

File: tools/test.sh (hunk 2 above)
Comment (thatux) [reply tools/test.sh]: done: changed the link

File: tools/test_modules/m34100.sh (hunk 3 above)
Comment (thatux) [reply tools/test_modules/m34100.sh]: I like them, makes logging much clearer, but I've removed them :)

**Comment by thatux** [comment]
All comments resolved :+1:
